### PR TITLE
Removed all the Statistics code.

### PR DESCRIPTION
**Pull request by paulp**
I think we should acknowledge this is the wrong way to approach
the issue. Statistics can be gathered in a way which does not
pollute the compiler codebase and is not rigidly tied into
statically generated logic, via instrumentation. Here are some
reasons why this code should go:
- it greatly obfuscates the logic of key methods
- it is untested
- it cannot even be enabled without recompiling
- it is inflexible in other ways
- it is almost certainly hurting performance

As to the last point, are we really that confident that a
difference like this in a hot method isn't going to have any
impact? I'm definitely not.

``` diff
-        : getstatic     #                 // Field util/Statistics$.MODULE$:Lutil/Statist
-        : invokevirtual #                 // Method util/Statistics$.canEnable:()Z
-        : ifeq
-       : getstatic     #                 // Field util/Statistics$.MODULE$:Lutil/Statisti
-       : getstatic     #                 // Field TypesStats$.MODULE$:LTypesStats$;
-       : invokevirtual #                 // Method TypesStats$.findMembersCount:()Lutil/S
-       : invokevirtual #                 // Method util/Statistics$.incCounter:(Lutil/St
-       : getstatic     #                 // Field util/Statistics$.MODULE$:Lutil/Statisti
-       : invokevirtual #                 // Method util/Statistics$.canEnable:()Z
-       : ifeq
-       : getstatic     #                 // Field util/Statistics$.MODULE$:Lutil/Statisti
-       : getstatic     #                 // Field TypesStats$.MODULE$:LTypesStats$;
-       : invokevirtual #                 // Method TypesStats$.typeOpsStack:()Lutil/Stati
-       : new           #                 // class Types$Type$$anonfun$
-       : dup
-       : invokespecial #                 // Method Types$Type$$anonfun$ ."<init>":(LTypes
-       : invokevirtual #                 // Method util/Statistics$.pushTimer:(Lutil/Stat
-       : goto
-       : aconst_null
-       : astore
-       : lload_3
-       : lstore
-       : lload_1
-       : ldc _w        #                 // long  l
+        : lload_3
+        : lstore
+        : lload_1
+        : ldc _w        #                 // long  l
```

### CODE REVIEW

**Comment by paulp** [comment]
Review by @gkossakowski

**Comment by som-snytt** [comment]
Sorry this isn't much of a review, but +1.  I tried using retronym's patched i/o counter, and it broke under Par-Test, or maybe I got the incantation wrong, my Latin is rusty.

**Comment by retronym** [comment]
Do you notice a performance change after the removal? In the current code, stats in hot methods should be guarded with`hotEnabled` (a compile time constant). A smaller change would be to make `canEnable` a compile time constant, too, and live with the fact that you need a special build to gather stats. Maybe not some a bit imposition for the peace of mind.

But I do agree about the intrusiveness of the code, and would like to see what's possible with external instrumentation.

If we stayed with in-code instrumentation, macros might offer a cleaner expression. But right now, I would oppose this, given that macros expansions currently end up losing debug line number information. That's a problem I need to solve in the context of async.

**Comment by paulp** [comment]
> Do you notice a performance change after the removal? In the current code, stats in hot methods should be guarded withhotEnabled (a compile time constant).

I haven't measured. Is the fact that it's guarded by a compile time constant going to bring it back below the MaxInlineSize threshold if the statistics code has taken it over it (which it will do more often than not)? I don't know, but I would be surprised if that was the case, and we can't afford that penalty.

But I think all this code should go even if there's no performance cost at all. I am here over the correctness cost.

**Comment by retronym** [comment]
Indeed `Inliners` runs before `dce` and `constant-optimization`.

**Comment by paulp** [comment]
Oh, you mean it's expunged by scalac. Only under -optimise. I guess if we ship it that way, it counts.

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -797,29 +789,18 @@ trait Types
...
-        (if (explainSwitch) explain("<:", isSubType, this, that)
-         else isSubType(this, that, AnyDepth))
-      if (Statistics.canEnable) Statistics.popTimer(typeOpsStack, start)
-      result
-    }
+    def stat_<:<(that: Type): Boolean = (
+         (this eq that)
+      || (if (explainSwitch) explain("<:", isSubType, this, that)
+          else isSubType(this, that, AnyDepth))
+    )
```
Comment (soc) [inline src/reflect/scala/reflect/internal/Types.scala:796]: `<:<` and `stat_<:<` are identical. Would it make sense to de-duplicate/remove one of them?

Comment (paulp) [inline src/reflect/scala/reflect/internal/Types.scala:796]: If consensus is that I can proceed with this, I'll get it.

**Comment by adriaanm** [comment]
I agree the statistics code is broken. However, unless there's a much more serious issue with it than I'm aware of, removal has to be contingent on having an alternative.

**Comment by paulp** [comment]
Can we define what "an alternative" would look like? Here is the output from -Ystatistics if I don't rebuild the compiler with different constants. For code which is only enabled if you rebuild the compiler differently, what would possess us to put it in master, where it impacts everybody, rather than in a place seen only by the two or three people who might ever do that?

I can certainly give you these statistics if these are something which it is important to give. (I seriously question that, but I will do it if it will let me proceed with the rest.)

```
% scalac3 -Ystatistics -d /tmp $(find ./src/library/scala/collection//mutable/ -name '*.scala')
Enabling statistics, measuring overhead = 59.1ns to 647.5ns per timer
*** Cumulative statistics at phase parser
#created tree nodes           : 37626
#created tree nodes by type   : 
#retained tree nodes          : 33282
#retained tree nodes by type  : Ident: 11190, Select: 5327, Apply: 2913, ValDef: 1574, DefDef: 1562, AppliedTypeTree: 1527, Literal: 1370, : 1337, TypeTree: 963, Block: 887, This: 743, New: 414, If: 367, TypeDef: 337, TypeBoundsTree: 328, Assign: 322, Super: 318, Template: 245, ClassDef: 191, : 179, PackageDef: 167, TypeApply: 167, SingletonTypeTree: 144, CaseDef: 139, Import: 104, Bind: 101, LabelDef: 80, Function: 55, Match: 54, Typed: 41, ModuleDef: 39, Throw: 38, ExistentialTypeTree: 24, CompoundTypeTree: 15, Return: 9, Annotated: 8, AssignOrNamedArg: 2, Try: 1
*** Cumulative statistics at phase typer
#symbols                      : 105821
#created tree nodes by type   : 
#unique types                 : 73966
#retained tree nodes          : 33489
#retained tree nodes by type  : Select: 6817, TypeTree: 5336, Ident: 3562, Apply: 3450, This: 3340, DefDef: 1939, ValDef: 1741, : 1498, Literal: 1379, Block: 961, TypeApply: 477, If: 368, New: 334, Super: 322, TypeDef: 313, Template: 242, Assign: 225, ClassDef: 191, PackageDef: 167, CaseDef: 166, Import: 104, Bind: 90, LabelDef: 80, Function: 71, Match: 67, Typed: 60, ApplyToImplicitArgs: 53, ModuleDef: 51, Throw: 40, ApplyImplicitView: 35, Return: 9, Try: 1, TypeBoundsTree: 0, : 0, AssignOrNamedArg: 0, ExistentialTypeTree: 0, SingletonTypeTree: 0, AppliedTypeTree: 0, CompoundTypeTree: 0, Annotated: 0
#created tree nodes           : 87431
*** Cumulative statistics at phase patmat
#symbols                      : 1992
#created tree nodes by type   : 
#unique types                 : 0
#retained tree nodes          : 35401
#retained tree nodes by type  : Select: 7075, TypeTree: 5639, Ident: 4124, Apply: 3785, This: 3340, DefDef: 1939, ValDef: 1928, Literal: 1401, : 1338, Block: 1126, TypeApply: 609, If: 498, New: 352, LabelDef: 326, Super: 322, TypeDef: 313, Template: 242, Assign: 231, ClassDef: 191, PackageDef: 167, Import: 104, Function: 71, Typed: 68, Throw: 58, ApplyToImplicitArgs: 53, ModuleDef: 51, ApplyImplicitView: 35, Return: 9, CaseDef: 4, Try: 1, Match: 1, TypeBoundsTree: 0, : 0, AssignOrNamedArg: 0, ExistentialTypeTree: 0, SingletonTypeTree: 0, AppliedTypeTree: 0, CompoundTypeTree: 0, Annotated: 0, Bind: 0
#created tree nodes           : 28
*** Cumulative statistics at phase erasure
#symbols                      : 1992
#created tree nodes by type   : 
#unique types                 : 0
#retained tree nodes          : 66202
#retained tree nodes by type  : TypeTree: 11874, Select: 11755, Apply: 10609, This: 6322, Ident: 5492, DefDef: 4705, ValDef: 3946, : 3838, Literal: 1513, TypeApply: 1498, Block: 1402, New: 576, If: 508, Template: 373, ClassDef: 373, LabelDef: 338, Super: 326, Assign: 231, Typed: 186, PackageDef: 167, Throw: 58, ApplyToImplicitArgs: 53, ApplyImplicitView: 35, Return: 9, ArrayValue: 9, CaseDef: 4, Try: 1, Match: 1, TypeDef: 0, TypeBoundsTree: 0, : 0, AssignOrNamedArg: 0, ExistentialTypeTree: 0, Function: 0, SingletonTypeTree: 0, Import: 0, ModuleDef: 0, AppliedTypeTree: 0, CompoundTypeTree: 0, Annotated: 0, Bind: 0
#created tree nodes           : 28
*** Cumulative statistics at phase cleanup
#symbols                      : 1992
#created tree nodes by type   : 
#unique types                 : 0
#retained tree nodes          : 119201
#retained tree nodes by type  : Select: 25837, TypeTree: 19822, Ident: 15857, Apply: 13912, This: 9889, : 8910, ValDef: 8906, DefDef: 8804, Literal: 1599, Block: 1382, TypeApply: 592, If: 584, New: 575, Assign: 550, Template: 373, ClassDef: 373, LabelDef: 338, Super: 310, Typed: 186, PackageDef: 167, Throw: 124, ApplyToImplicitArgs: 52, ApplyImplicitView: 35, Return: 9, ArrayValue: 9, CaseDef: 4, Try: 1, Match: 1, TypeDef: 0, TypeBoundsTree: 0, : 0, AssignOrNamedArg: 0, ExistentialTypeTree: 0, Function: 0, SingletonTypeTree: 0, Import: 0, ModuleDef: 0, AppliedTypeTree: 0, CompoundTypeTree: 0, Annotated: 0, Bind: 0
#created tree nodes           : 28
```

**Comment by adriaanm** [comment]
> what would possess us to put it in master, where it impacts everybody, rather than in a place seen only by the two or three people who might ever do that?

Because that gives us another branch to build for every commit in every PR, every night,... to make sure it doesn't break. We are not yet in a position where that is cheap enough to do. 

Could we use the elision mechanism (or the macro equivalent) that still compiles the code but makes sure it doesn't get run/variables don't get emitted? But then again, are we ready to start using macros ourselves?

Honestly, I see a lot of moving parts here that I'm not sure should start moving right now.

**Comment by paulp** [comment]
> Because that gives us another branch to build for every commit in every PR, every night,... to make sure it doesn't break.

That's kind of my question - what is the value of "it" here? What does it mean for it to break? If you mean "it still compiles", that is a very superficial test of whether something is broken.

And my interest is more in the deobfuscation of thsee methods than it is in any detail of the generated bytecode, so any change which fails to free them from their present syntactic obliteration is unappealing. We must have more eyes on code; these methods are the most important to have eyes on; the noise is a terribly effective eye-repellent.

**Comment by adriaanm** [comment]
The noise is a compelling argument. It's unfortunate the bar currently only is at "it compiles". I'd like us to get to "it makes pretty charts and detects regressions". "It can be used to quickly diagnose a problem." If we remove "it" from the codebase, we're taking a step back from these goals. Aren't we?

**Comment by retronym** [comment]
We could certainly make a substantial incremental improvement on the syntax. If we simplify the design space by admitting that this one needs to use a purpose built compiler for statistics gathering, we could get rid of `if`-s at all the call sites in favour of `@elide`.

``` diff
-val start = if (Statistics.canEnable) Statistics.startTimer(patmatAnaVarEq) else null
+val start = patmatAnaVarEq.start
```

``` diff
-if (Statistics.canEnable) Statistics.incCounter(plausiblyCompatibleImplicits)
+plausiblyCompatibleImplicits.inc()
```

We should consider the (important!) aesthetic argument by weighing that sort of code against no-statistics-at-all code.

As to alternatives, I usually find YourKit profiles more useful than the compiler generated timings. (Although, sadly, neither have led me to any Eureka! moments in identifying areas for improvement, after spending way too long poring over them.)

We can make YourKit aware of compiler phases, by advancing the [object generation](http://www.yourkit.com/docs/80/api/index.html) after each phase. This lets you determine if a particular phase is allocation heavy. We could also capture separate CPU snapshots for each profile. I'm pretty sure Paul has done some of this in the past. It's probably not _that_ big a win, but at least it's possible.

The instrumented counters are still pretty useful. I'd like to keep them.

I've also noticed that the current approach can omit useful spans based on assumptions on the part of the instrumenter. Right now, `asSeenFrom` stats are not reported post-typer, but I found that 10% of the time compiling the collections library comes from this operation, as used in overriding pairs computations in refchecks, and later in bridge generation.

**Comment by gkossakowski** [comment]
I never found calls to Statistics that disturbing. They are very easy to spot and very regular so it's ignoring them when reading code is not really a problem.

My arguments for removing current Statistics implementation would be:
- they do not work with multiple Globals as in sbt or IDE; we have one static object for Statistics that is referred and mutated from inside of the cake, that's a very bad design that leads to incorrect results reported
- they are hard to adjust to specific problem you are trying to detect because you have to rebuild the whole compiler and this option is meant to be used by other people to provide us (compiler hackers) useful information when diagnosing problems with code we don't have an access to. Shipping custom built of the compiler is very often not an option

Those problems inspired me to experiment with AspectJ for instrumenting the compiler. My experiments show that this is a feasible option that has _potential_ to address both of problems outlined above. Also, it would free us from polluting compiler code with instrumentation code.

However, `-Ystatistics` does work for batch compilation (with one global) and does print useful results right now and we don't have AspectJ-based replacement for it yet. If you would like to work on the replacement I'm happy to share my thoughts and help by doing code reviews.

Therefore I'd like to keep things as-is for now. I'm open for more arguments supporting the removal. For example, if we can get convincing argument that this code really hurts performance in significant way we can revisit this discussion.

**Comment by paulp** [comment]
You see right through this stuff? More power to you; but you had best realize other people don't.

```
  def inferImplicit(tree: Tree, pt: Type, reportAmbiguous: Boolean, isView: Boolean, context: Context, saveAmbiguousDivergent: Boolean, pos: Position): SearchResult = {
    printInference("[infer %s] %s with pt=%s in %s".format(
      if (isView) "view" else "implicit",
      tree, pt, context.owner.enclClass)
    )
    printTyping(
      ptBlock("infer implicit" + (if (isView) " view" else ""),
        "tree"        -> tree,
        "pt"          -> pt,
        "undetparams" -> context.outer.undetparams
      )
    )
    indentTyping()

    val rawTypeStart    = if (Statistics.canEnable) Statistics.startCounter(rawTypeImpl) else null
    val findMemberStart = if (Statistics.canEnable) Statistics.startCounter(findMemberImpl) else null
    val subtypeStart    = if (Statistics.canEnable) Statistics.startCounter(subtypeImpl) else null
    val start           = if (Statistics.canEnable) Statistics.startTimer(implicitNanos) else null
    if (printInfers && !tree.isEmpty && !context.undetparams.isEmpty)
      printTyping("typing implicit: %s %s".format(tree, context.undetparamsString))
    val implicitSearchContext = context.makeImplicit(reportAmbiguous)
    val result = new ImplicitSearch(tree, pt, isView, implicitSearchContext, pos).bestImplicit
    if (saveAmbiguousDivergent && implicitSearchContext.hasErrors) {
      context.updateBuffer(implicitSearchContext.errBuffer.filter(err => err.kind == ErrorKinds.Ambiguous || err.kind == ErrorKinds.Divergent))
      debuglog("update buffer: " + implicitSearchContext.errBuffer)
    }
    printInference("[infer implicit] inferred " + result)
    context.undetparams = context.undetparams filterNot result.subst.from.contains

    if (Statistics.canEnable) Statistics.stopTimer(implicitNanos, start)
    if (Statistics.canEnable) Statistics.stopCounter(rawTypeImpl, rawTypeStart)
    if (Statistics.canEnable) Statistics.stopCounter(findMemberImpl, findMemberStart)
    if (Statistics.canEnable) Statistics.stopCounter(subtypeImpl, subtypeStart)
    deindentTyping()
    printTyping("Implicit search yielded: "+ result)
    result
  }
```

**Comment by paulp** [comment]
> Those problems inspired me to experiment with AspectJ for instrumenting the compiler. My experiments show that this is a feasible option that has potential to address both of problems outlined above. Also, it would free us from polluting compiler code with instrumentation code.

Anticipating the risk people would make me come up with a replacement, even if I question the legitimacy of having to replace something with the aforementioned qualities, I already resumed hacking on this. I have easily customizable invocation counts now; to count time I want to translate into java the method I used for tree timings, which offered those pretty ascii graphs.

```
 1610/141  66819: package scalaz                                       (line 1)
 1468/0    | 68842: object Equal extends EqualL...                     (line 60)
 1468/682  | | 68847: Template(value <local Equal>)                    (line 60)
  229/0    | | | 70921: implicit def TreeEqual[A: E...                 (line 189)
  229/0    | | | | 70937: equal[Tree[A]]((a1, a2) =>                   (line 190)
  229/0    | | | | | 70943: Function(value $anonfun)                   (line 190)
  229/11   | | | | | | 70950: && i[Equal[Iterable[Tree[A]...           (line 192)
  217/0    | | | | | | | 70967: Apply                                  (line 192)
  217/0    | | | | | | | | 70968: Select(TypeApply, equal) equal       (line 192)
  217/23   | | | | | | | | | 70969: TypeApply                          (line 192)
  193/0    | | | | | | | | | | 70987: ApplyToImplicitArgs(method implicitly) (line 192)
  193/11   | | | | | | | | | | | 70989: TypeApply(method TraversableEqual) (line 192)
  181/3    | | | | | | | | | | | | 70994: ApplyToImplicitArgs(method TraversableEqual) (line 192)
  178/     | | | | | | | | | | | | | 70998: TypeApply(method TreeEqual) (line 192)
           | | | | | | | | | | | | | \--70998
           | | | | | | | | | | | | \--70994
           | | | | | | | | | | | \--70989
           | | | | | | | | | | \--70987
           | | | | | | | | | \--70969
           | | | | | | | | \--70968
           | | | | | | | \--70967
           | | | | | | \--70950
           | | | | | \--70943
           | | | | \--70937
           | | | \--70921
  556/0    | | | 71005: implicit def TreeLocEqual[A...                 (line 194)
  555/0    | | | | 71021: equal[TreeLoc[A]]((a1, a2) =>                (line 195)
  555/0    | | | | | 71027: Function(value $anonfun)                   (line 195)
  555/0    | | | | | | 71034: && a1.lefts ≟ a2.lefts && a...           (line 197)
  152/0    | | | | | | | 71035: Select(Apply, &&) &&                   (line 197)
  152/     | | | | | | | | 71036: Apply                                (line 197)
           | | | | | | | | \--71036
           | | | | | | | \--71035
  403/237  | | | | | | | 71136: Apply                                  (line 197)
  165/0    | | | | | | | | 71214: ApplyToImplicitArgs(method ≟)        (line 197)
  165/     | | | | | | | | | 71216: TypeApply(method TraversableEqual) (line 197)
           | | | | | | | | | \--71216
           | | | | | | | | \--71214
           | | | | | | | \--71136
           | | | | | | \--71034
           | | | | | \--71027
           | | | | \--71021
           | | | \--71005
           | | \--68847
           | \--68842
           \--66819
```

**Comment by paulp** [comment]
Here's the kind of interesting info one can acquire by virtue of the instrumentation approach, where there's zero chance one would pursue it for all three compiler source bases. These are the number of calls to particular methods when compiling src/swing with 2.9.3, 2.10.1, and master. (Don't take the numbers too seriously, since the 2.11 build wasn't compiled under -optimise like the release versions.)

``````
          assert  2.9.3:   12101422  public void scala.Predef..assert(boolean)
          assert 2.10.1:    8699268  public void scala.Predef..assert(boolean)
          assert   2.11:    7826882  public void scala.Predef..assert(boolean)
        canEqual  2.9.3:     589754  public boolean scala.tools.nsc.symtab.Types.TypeRef.canEqual(java.lang.Object)
        canEqual 2.10.1:     344481  public boolean scala.reflect.internal.Types.TypeRef.canEqual(java.lang.Object)
        canEqual   2.11:     344151  public boolean scala.reflect.internal.Types.TypeRef.canEqual(java.lang.Object)
           clone  2.9.3:      12003  public final scala.tools.nsc.symtab.Symbols.Symbol scala.tools.nsc.symtab.Symbols..anonfun.cloneSymbols
           clone 2.10.1:      44483  public final scala.reflect.internal.Symbols.Symbol scala.reflect.internal.Symbols..anonfun.cloneSymbols
           clone   2.11:      45047  public final scala.reflect.internal.Symbols.Symbol scala.reflect.internal.Symbols..anonfun.cloneSymbols
        conforms  2.9.3:     164080  public scala.Predef..less.colon.less scala.Predef..conforms()
        conforms 2.10.1:      26536  public scala.Predef..less.colon.less scala.Predef..conforms()
        conforms   2.11:      25260  public scala.Predef..less.colon.less scala.Predef..conforms()
            copy  2.9.3:     115156  public void scala.Array..copy(java.lang.Object, int, java.lang.Object, int, int)
            copy 2.10.1:      44090  public void scala.Array..copy(java.lang.Object, int, java.lang.Object, int, int)
            copy   2.11:      44676  public void scala.Array..copy(java.lang.Object, int, java.lang.Object, int, int)```
``````

**Comment by retronym** [comment]
I'm going to close this for now (to ease congestion, not as endorsement for the status quo.)

**Pull request closed without merging**